client/admin.go

```diff
@@ -59,31 +65,91 @@ func CreateAndInitTree(ctx context.Context, req *trillian.CreateTreeRequest, adm
 		return nil, err
 	}
 
-	if tree.TreeType == trillian.TreeType_MAP {
-		b.Reset()
-		err := b.Retry(ctx, func() error {
-			glog.Infof("Initialising Map %x...", tree.TreeId)
-			// Now, if it's a Map, initialise it.
-			req := &trillian.InitMapRequest{MapId: tree.TreeId}
-			resp, err := mapClient.InitMap(ctx, req)
-			if err != nil {
-				switch s, ok := status.FromError(err); {
-				case ok && s.Code() == codes.Unavailable:
-					glog.Errorf("Map server unavailable: %v", err)
-					return err
-				case ok && s.Code() == codes.AlreadyExists:
-					glog.Warningf("Bizarrely, the just-created Map (%x) is already initialised!: %v", tree.TreeId, err)
-					return err
-				}
-				return fmt.Errorf("failed to InitMap(%+v): %T %v", req, err, err)
+	switch tree.TreeType {
+	case trillian.TreeType_MAP:
+		err = InitMap(ctx, tree, mapClient)
+	case trillian.TreeType_LOG:
+		err = InitLog(ctx, tree, logClient)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return tree, nil
+}
+
+// InitMap initialises a freshly created Map tree.
+func InitMap(ctx context.Context, tree *trillian.Tree, mapClient trillian.TrillianMapClient) error {
+	if tree.TreeType != trillian.TreeType_MAP {
+		return fmt.Errorf("InitMap called with tree of type %v", tree.TreeType)
+	}
+
+	b := &backoff.Backoff{
+		Min:    100 * time.Millisecond,
+		Max:    10 * time.Second,
+		Factor: 2,
+		Jitter: true,
+	}
+
+	err := b.Retry(ctx, func() error {
+		glog.Infof("Initialising Map %x...", tree.TreeId)
+		req := &trillian.InitMapRequest{MapId: tree.TreeId}
+		resp, err := mapClient.InitMap(ctx, req)
+		if err != nil {
+			switch s, ok := status.FromError(err); {
+			case ok && s.Code() == codes.Unavailable:
+				glog.Errorf("Map server unavailable: %v", err)
+				return err
+			case ok && s.Code() == codes.AlreadyExists:
+				glog.Warningf("Bizarrely, the just-created Map (%x) is already initialised!: %v", tree.TreeId, err)
+				return err
 			}
-			glog.Infof("Initialised Map (%x) with new SignedMapRoot:\n%+v", tree.TreeId, resp.Created)
+			return fmt.Errorf("failed to InitMap(%+v): %T %v", req, err, err)
+		}
+		glog.Infof("Initialised Map (%x) with new SignedMapRoot:\n%+v", tree.TreeId, resp.Created)
 
-			return nil
-		})
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// InitLog initialises a freshly created Log tree.
+func InitLog(ctx context.Context, tree *trillian.Tree, logClient trillian.TrillianLogClient) error {
```

Comment:
InitLog and InitMap are very similar.  Would be good if they could share code but that's probably not easy because of different client types etc.

Reply:
Yeah - I thought of that but it'll end up just looking quite like `Backoff` with a large function being passed in, or some hair ball of strings and bits... didn't seem worth it.